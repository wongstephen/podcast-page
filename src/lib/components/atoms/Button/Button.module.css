.container {
  --outline--width: 1px;
  --outline--offset: 3px;

  --border-width: 1px;

  --transition-time: 525ms; /* NO SPEC: taken from https://aircall.io/brand/ */
  --transition-property: ease-in-out;

  font-family: var(--font-primary);

  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;

  border-radius: var(--border-radius-sm);
  border: var(--border-width) solid transparent;
  outline: none;

  cursor: pointer;
  transition: color var(--transition-time) var(--transition-property),
    background var(--transition-time) var(--transition-property),
    border var(--transition-time) var(--transition-property);

  /* Appearance */
  &[data-appearance='primary'] {
    background-color: var(--color-green-500);
    color: var(--color-white);
    font-weight: var(--font-weight-semibold);

    &:hover {
      /* NO SPEC */
      background-color: var(--color-green-600);
    }

    &:hover:active {
      background-color: var(--color-green-600);
    }
  }

  &[data-appearance='outline'] {
    background-color: transparent;
    border-color: var(--color-green-500);
    color: var(--color-green-500);
    font-weight: var(--font-weight-semibold);

    &:hover {
      border-color: var(--color-green-600);
      color: var(--color-green-600);
    }

    &:hover:active {
      border-color: var(--color-green-600);
      color: var(--color-green-600);
    }
  }

  &[data-appearance='outline-secondary'] {
    background-color: transparent;
    border-color: var(--color-gray-700);
    color: var(--color-gray-600);
    font-weight: var(--font-weight-medium);

    &:hover {
      border-color: var(--color-gray-300);
    }

    &:hover:active {
      border-color: var(--color-gray-300);
    }
  }

  &[data-appearance='transparent'] {
    background-color: transparent;
    border-color: transparent;
    color: var(--color-green-600);
  }

  /* Disabled */
  &[data-disabled='true'] {
    background-color: var(--disabled-button);
    border-color: transparent;
    color: var(--color-white);
    cursor: not-allowed;
    pointer-events: none;
  }

  /* Button Sizes */
  &[data-size='small'] {
    font-size: var(--font-size-sm); /* 14px */
    padding: 8px 19px;
  }

  &[data-size='medium'] {
    font-size: var(--font-size-md); /* 16px */
    padding: 12px 19px;
  }

  /* Focus outline */
  &:focus-visible::before,
  &:focus-visible::after {
    content: '';
    position: absolute;
    inset: 0;
    border-radius: calc(var(--border-radius-sm) - var(--outline--offset));
  }

  &:focus-visible::before {
    outline: var(--outline--width) solid var(--background);
    outline-offset: var(--outline--offset);
  }
  &:focus-visible::after {
    outline: var(--outline--width) solid var(--color-green-500);
    outline-offset: calc(var(--outline--offset) + var(--outline--width));
  }
}
