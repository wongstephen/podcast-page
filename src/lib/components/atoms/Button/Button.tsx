import React from 'react';
import { Slot } from '@radix-ui/react-slot';

import styles from './Button.module.css';
import { cn } from '@/lib/utilities/cn';

type ButtonSizes = 'small' | 'medium' | 'none';

type ButtonAppearance =
  | 'primary'
  | 'outline'
  | 'outline-secondary'
  | 'transparent';

export type ButtonProps = {
  /**
   * A button can have different appearances and passed custom content.
   * - 'primary (default)': Primary button action emphasis.
   * - 'outline': Button removing the background.
   * - 'outline-secondary': A secondary outline button.
   * @default 'primary'
   */
  appearance?: ButtonAppearance;

  /**
   * Disables the button, preventing any interaction.
   *
   * @default false
   */
  disabled?: boolean;

  /**
   * Button size options
   *
   * @default 'medium'
   */
  size?: ButtonSizes;
  /**
   * Slot for the button content. When true, the styling will be applied to the child element.
   *
   * @default: false
   */
  asChild?: boolean;
} & React.ComponentProps<'button'>;

const Button = ({
  appearance = 'primary',
  className,
  asChild = false,
  children,
  disabled,
  size = 'medium',
  ...props
}: ButtonProps): React.JSX.Element => {
  const Comp = asChild ? Slot : 'button';

  return (
    <Comp
      className={cn(styles.container, className)}
      data-appearance={appearance}
      data-disabled={disabled}
      data-size={size}
      disabled={disabled}
      {...props}
    >
      {children}
    </Comp>
  );
};

Button.displayName = 'Button';

export default Button;
