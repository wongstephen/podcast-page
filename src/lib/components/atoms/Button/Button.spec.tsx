import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import Button from './Button';

describe('Button Component', () => {
  it('matches snapshot', () => {
    const { container } = render(<Button>Snapshot Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it('renders the button with the correct text', () => {
    render(<Button>Click Me</Button>);
    const buttonElement = screen.getByText('Click Me');
    expect(buttonElement).toBeInTheDocument();
  });

  it('applies the correct appearance styles', () => {
    render(<Button data-appearance="primary">Primary Button</Button>);
    const buttonElement = screen.getByText('Primary Button');
    expect(buttonElement).toHaveClass('container');
    expect(buttonElement).toHaveAttribute('data-appearance', 'primary');
  });

  it('handles click events', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click Me</Button>);
    const buttonElement = screen.getByText('Click Me');
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('disables the button when disabled is true', () => {
    render(<Button disabled>Disabled Button</Button>);
    const buttonElement = screen.getByText('Disabled Button');
    expect(buttonElement).toBeDisabled();
  });
});
