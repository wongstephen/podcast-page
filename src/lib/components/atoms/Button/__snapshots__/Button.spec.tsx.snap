// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component matches snapshot 1`] = `
<div>
  <button
    class="container"
    data-appearance="primary"
    data-size="medium"
  >
    Snapshot Button
  </button>
</div>
`;
